ballista/rust/core/src/serde/logical_plan/mod.rs

```diff
@@ -1016,6 +1016,55 @@ mod roundtrip_tests {
                 "this is only a test object store".to_string(),
             ))
         }
+
+        fn file_writer(
+            &self,
+            _path: &str,
+        ) -> datafusion_data_access::Result<Arc<dyn ObjectWriter>> {
+            unimplemented!();
+        }
+
+        async fn create_dir(
```

Comment:
I was curious what this corresponded to given that object stores don't have a concept of a directory. It would appear that at least in the case of the S3 implementation, it creates empty objects as pretend directories - https://github.com/apache/arrow/blob/master/cpp/src/arrow/filesystem/s3fs.cc

I'm not sure how I feel about this

Reply:
That's a good point. I didn't know that was the behavior in C++ Arrow. I started a sibling PR to implement these methods for S3 in https://github.com/datafusion-contrib/datafusion-objectstore-s3/pull/54, so I might prototype some more there and see if that's necessary in any way.

Reply:
It looks like fsspec implementation [doesn't create empty objects as directories](https://github.com/fsspec/s3fs/blob/c4fb41f7cc2f2aede6bbb7755096c38b9e4cc553/s3fs/core.py#L730-L761), so I think it is sensible to do the same. 

It's not totally a no-op though for cases like S3. You might want to have `create_dir` create a bucket if it doesn't exist (though I'd like for there to be an option to turn off bucket creation). And it should probably check that there isn't an existing file at the path you are trying to create a directory at.

I think what I will do is document the expected behavior and create a generic test that can be run on an arbitrary filesystem to help implementation make sure they follow expected behavior.